okhttp-testing-support/src/main/java/okhttp3/OkHttpClientTestRule.kt

```diff
@@ -55,25 +55,52 @@ class OkHttpClientTestRule : TestRule {
         acquireClient()
         try {
           base.evaluate()
+          logEventsIfFlaky(description.testClass.canonicalName, description.methodName)
+        } catch (t: Throwable) {
+          logEventsOnFailure(t)
+          throw t
         } finally {
           ensureAllConnectionsReleased()
           releaseClient()
         }
       }
 
       private fun acquireClient() {
-        prototype = prototypes.poll() ?: OkHttpClient.Builder()
-            .dns(SINGLE_INET_ADDRESS_DNS) // Prevent unexpected fallback addresses.
-            .build()
+        prototype = prototypes.poll() ?: freshClient()
+        clientEventsMap[prototype!!] = mutableListOf()
       }
 
       private fun releaseClient() {
         prototypes.push(prototype)
+        clientEventsMap.remove(prototype)
         prototype = null
       }
     }
   }
 
+  private fun logEventsIfFlaky(testClass: String, testMethod: String) {
+    if (flakyTests.contains(testClass) || flakyTests.contains("$testClass#$testMethod")) {
+      logEvents()
+    }
+  }
+
+  private fun logEventsOnFailure(t: Throwable) {
```

Comment:
why not just call logEvents ?

Reply:
I think I wanted this to also account for flaky tests... I forget :)